File: frontend/src/services/CharacterParser.js
// JavaScript Character Parser - Equivalent to Python version
export class CharacterParser {
  constructor() {
    this.sectionPatterns = [
      /^[A-ZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀĀĂĄĆČĎĐĒĔĖĘĚĜĞĠĢĤĦĪĬĮİĲĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŪŬŮŰŲŴŶŸŹŻŽ\s]+$/,
      /^[A-ZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀĀĂĄĆČĎĐĒĔĖĘĚĜĞĠĢĤĦĪĬĮİĲĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŪŬŮŰŲŴŶŸŹŻŽ\s].*[A-ZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀĀĂĄĆČĎĐĒĔĖĘĚĜĞĠĢĤĦĪĬĮİĲĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŪŬŮŰŲŴŶŸŹŻŽ]$/
    ];
    
    this.decorativePatterns = [
      /^={3,}/,  // ===
      /^-{3,}/,  // ---
      /^_{3,}/,  // ___
      /^#{3,}/,  // ###
      /^\*{3,}/, // ***
    ];
  }

  isTopicLine(line) {
    const trimmed = line.trim();
    if (!trimmed) return false;
    // Check if line ends with colon and is not too long
    return trimmed.endsWith(':') && trimmed.length > 3 && trimmed.length < 100;
  }

  isSectionHeader(line) {
    const trimmed = line.trim();
    if (!trimmed) return false;
    
    // Check for decorative patterns
    for (const pattern of this.decorativePatterns) {
      if (pattern.test(trimmed)) return false;
    }
    
    // Check for topic lines (they are not section headers)
    if (this.isTopicLine(trimmed)) return false;
    
    // Check for section patterns
    return this.sectionPatterns.some(pattern => pattern.test(trimmed));
  }

  isDecorativeLine(line) {
    const trimmed = line.trim();
    return this.decorativePatterns.some(pattern => pattern.test(trimmed));
  }

  extractCharacterName(content) {
    const lines = content.split('\n');
    
    for (const line of lines) {
      const trimmed = line.trim();
      
      // Look for "NOME DO PERSONAGEM:" pattern
      if (trimmed.toUpperCase().includes('NOME DO PERSONAGEM:')) {
        const parts = trimmed.split(':', 2);
        if (parts.length > 1) {
          return parts[1].trim();
        }
      }
      
      // Look for "PERFIL DE PERSONAGEM -" pattern
      if (trimmed.toUpperCase().includes('PERFIL DE PERSONAGEM -')) {
        const parts = trimmed.split('-', 2);
        if (parts.length > 1) {
          return parts[1].trim();
        }
      }
    }
    
    return "Nome não encontrado";
  }

  parseContentIntoSections(content) {
    const lines = content.split('\n');
    const sections = [];
    let currentSection = null;
    let currentItems = [];
    let currentTopic = null;
    let currentSubitems = [];
    
    for (const line of lines) {
      const trimmed = line.trim();
      
      // Skip empty lines and decorative lines
      if (!trimmed || this.isDecorativeLine(trimmed)) {
        continue;
      }
      
      // Check if it's a section header
      if (this.isSectionHeader(trimmed)) {
        // Save previous topic if exists
        if (currentTopic && currentSubitems.length > 0) {
          currentItems.push({
            subtitle: currentTopic,
            text: currentSubitems.join('\n')
          });
          currentTopic = null;
          currentSubitems = [];
        }
        
        // Save previous section if exists
        if (currentSection && currentItems.length > 0) {
          sections.push({
            title: currentSection,
            content: currentItems
          });
        }
        
        // Start new section
        currentSection = trimmed;
        currentItems = [];
        continue;
      }
      
      // Process content within section
      if (currentSection) {
        // Check if it's a topic line (ends with colon)
        if (this.isTopicLine(trimmed)) {
          // Save previous topic if exists
          if (currentTopic && currentSubitems.length > 0) {
            currentItems.push({
              subtitle: currentTopic,
              text: currentSubitems.join('\n')
            });
          }
          
          // Start new topic
          currentTopic = trimmed.slice(0, -1); // Remove colon
          currentSubitems = [];
          continue;
        }
        
        // If we have a current topic, add to subitems
        if (currentTopic) {
          // Check if line contains a colon (key-value pair within topic)
          if (trimmed.includes(':') && !trimmed.startsWith('-') && !trimmed.startsWith('*')) {
            const parts = trimmed.split(':', 2);
            if (parts.length === 2) {
              const key = parts[0].trim();
              const value = parts[1].trim();
              if (key && value) {
                currentSubitems.push(`**${key}**: ${value}`);
                continue;
              }
            }
          }
          
          // Check if it's a numbered or bulleted list item
          if (/^[\d\-\*]\s*\.?\s*/.test(trimmed)) {
            const cleaned = trimmed.replace(/^[\d\-\*]\s*\.?\s*/, '').trim();
            if (cleaned) {
              currentSubitems.push(`• ${cleaned}`);
            }
            continue;
          }
          
          // Regular line under topic
          if (trimmed.length > 3) {
            currentSubitems.push(trimmed);
          }
          continue;
        }
        
        // No current topic, process as regular item
        // Check if line contains a colon (key-value pair)
        if (trimmed.includes(':') && !trimmed.startsWith('-') && !trimmed.startsWith('*')) {
          const parts = trimmed.split(':', 2);
          if (parts.length === 2) {
            const key = parts[0].trim();
            const value = parts[1].trim();
            if (key && value) {
              currentItems.push({
                subtitle: key,
                text: value
              });
              continue;
            }
          }
        }
        
        // Check if it's a numbered or bulleted list item
        if (/^[\d\-\*]\s*\.?\s*/.test(trimmed)) {
          const cleaned = trimmed.replace(/^[\d\-\*]\s*\.?\s*/, '').trim();
          if (cleaned) {
            currentItems.push({
              text: cleaned
            });
          }
          continue;
        }
        
        // Regular paragraph
        if (trimmed.length > 10) {
          currentItems.push({
            text: trimmed
          });
        }
      }
    }
    
    // Don't forget the last topic and section
    if (currentTopic && currentSubitems.length > 0) {
      currentItems.push({
        subtitle: currentTopic,
        text: currentSubitems.join('\n')
      });
    }
    
    if (currentSection && currentItems.length > 0) {
      sections.push({
        title: currentSection,
        content: currentItems
      });
    }
    
    return sections;
  }

  parseCharacterFile(fileName, content) {
    try {
      const characterName = this.extractCharacterName(content);
      const sections = this.parseContentIntoSections(content);
      
      // Extract basic info from first section if available
      let title = "Personagem";
      if (sections.length > 0 && sections[0].title.toUpperCase().includes('INFORMAÇÕES BÁSICAS')) {
        for (const item of sections[0].content) {
          if (item.subtitle && item.subtitle.toLowerCase() === 'tipo') {
            title = item.text;
            break;
          }
        }
      }
      
      return {
        id: fileName.replace('.txt', ''),
        name: characterName,
        title: title,
        sections: sections,
        file_path: fileName,
        word_count: content.split(/\s+/).length,
        section_count: sections.length
      };
    } catch (error) {
      console.error(`Error parsing file ${fileName}:`, error);
      return null;
    }
  }

  searchCharacters(characters, query) {
    if (!query) return characters;
    
    const queryLower = query.toLowerCase();
    const filteredCharacters = [];
    
    for (const character of characters) {
      // Search in name and title
      if (character.name.toLowerCase().includes(queryLower) || 
          character.title.toLowerCase().includes(queryLower)) {
        filteredCharacters.push(character);
        continue;
      }
      
      // Search in section content
      let foundInContent = false;
      for (const section of character.sections) {
        if (section.title.toLowerCase().includes(queryLower)) {
          foundInContent = true;
          break;
        }
        for (const item of section.content) {
          if (item.text && item.text.toLowerCase().includes(queryLower)) {
            foundInContent = true;
            break;
          }
          if (item.subtitle && item.subtitle.toLowerCase().includes(queryLower)) {
            foundInContent = true;
            break;
          }
        }
        if (foundInContent) break;
      }
      
      if (foundInContent) {
        filteredCharacters.push(character);
      }
    }
    
    return filteredCharacters;
  }

  getCharacterMetrics(characters) {
    if (!characters || characters.length === 0) return {};
    
    const totalCharacters = characters.length;
    const totalSections = characters.reduce((sum, char) => sum + char.section_count, 0);
    const totalWords = characters.reduce((sum, char) => sum + char.word_count, 0);
    
    // Section analysis
    const sectionTitles = [];
    for (const char of characters) {
      sectionTitles.push(...char.sections.map(section => section.title));
    }
    
    const sectionFrequency = this.getFrequency(sectionTitles);
    
    // Character title analysis
    const characterTitles = characters.map(char => char.title);
    const titleFrequency = this.getFrequency(characterTitles);
    
    // Common words analysis
    const allText = [];
    for (const char of characters) {
      for (const section of char.sections) {
        for (const item of section.content) {
          if (item.text) allText.push(item.text);
          if (item.subtitle) allText.push(item.subtitle);
        }
      }
    }
    
    const stopWords = new Set(['de', 'da', 'do', 'das', 'dos', 'a', 'o', 'e', 'em', 'para', 'por', 'com', 'um', 'uma', 'que', 'não', 'é', 'se', 'na', 'no', 'ou', 'como', 'mais', 'sua', 'seu', 'suas', 'seus', 'os', 'as', 'ao', 'aos', 'à', 'às', 'pelo', 'pela', 'pelos', 'pelas', 'sobre', 'sob', 'entre', 'através', 'durante', 'depois', 'antes', 'já', 'ainda', 'também', 'sempre', 'nunca', 'todos', 'todas', 'cada', 'outro', 'outra', 'outros', 'outras', 'mesmo', 'mesma', 'mesmos', 'mesmas', 'muito', 'muita', 'muitos', 'muitas', 'pouco', 'pouca', 'poucos', 'poucas', 'grande', 'grandes', 'pequeno', 'pequena', 'pequenos', 'pequenas', 'primeiro', 'primeira', 'primeiros', 'primeiras', 'último', 'última', 'últimos', 'últimas', 'bem', 'mal', 'melhor', 'pior', 'maior', 'menor', 'pode', 'podem', 'deve', 'devem', 'tem', 'têm', 'há', 'havia', 'houve', 'será', 'serão', 'foi', 'foram', 'seja', 'sejam', 'sendo', 'sido', 'estar', 'estão', 'está', 'esteja', 'estejam', 'estando', 'estado', 'ter', 'tendo', 'tido', 'fazer', 'faz', 'fazem', 'feito', 'feitos', 'feita', 'feitas', 'fazendo', 'dar', 'dá', 'dão', 'dado', 'dados', 'dada', 'dadas', 'dando', 'ver', 'vê', 'veem', 'visto', 'vista', 'vistos', 'vistas', 'vendo', 'ir', 'vai', 'vão', 'ido', 'ida', 'idos', 'idas', 'indo', 'vir', 'vem', 'vêm', 'vindo', 'vindo']);
    
    const wordFreq = new Map();
    for (const text of allText) {
      const words = text.toLowerCase().match(/\b\w+\b/g) || [];
      for (const word of words) {
        if (word.length > 3 && !stopWords.has(word)) {
          wordFreq.set(word, (wordFreq.get(word) || 0) + 1);
        }
      }
    }
    
    const commonThemes = Array.from(wordFreq.entries())
      .filter(([word, freq]) => freq > 1)
      .sort(([,a], [,b]) => b - a)
      .slice(0, 20)
      .map(([word, frequency]) => ({ word, frequency }));
    
    return {
      total_characters: totalCharacters,
      total_sections: totalSections,
      total_words: totalWords,
      avg_sections_per_character: Math.round((totalSections / totalCharacters) * 100) / 100,
      avg_words_per_character: Math.round((totalWords / totalCharacters) * 100) / 100,
      most_common_sections: Object.fromEntries(Object.entries(sectionFrequency).slice(0, 10)),
      character_types: titleFrequency,
      common_themes: commonThemes,
      section_diversity: Object.keys(sectionFrequency).length,
      unique_sections: Object.keys(sectionFrequency)
    };
  }

  getFrequency(items) {
    const frequency = {};
    for (const item of items) {
      frequency[item] = (frequency[item] || 0) + 1;
    }
    return frequency;
  }

  findSimilarities(char1, char2) {
    if (!char1 || !char2) return {};
    
    const sections1 = new Set(char1.sections.map(s => s.title));
    const sections2 = new Set(char2.sections.map(s => s.title));
    
    const commonSections = [...sections1].filter(s => sections2.has(s));
    const uniqueToChar1 = [...sections1].filter(s => !sections2.has(s));
    const uniqueToChar2 = [...sections2].filter(s => !sections1.has(s));
    
    const allSections = new Set([...sections1, ...sections2]);
    const overallSimilarity = commonSections.length / allSections.size;
    
    return {
      character1: char1.name,
      character2: char2.name,
      common_sections: commonSections,
      unique_to_char1: uniqueToChar1,
      unique_to_char2: uniqueToChar2,
      overall_similarity: Math.round(overallSimilarity * 1000) / 1000
    };
  }
}

File: backend/services/character_parser.py
import re
import os
from typing import List, Dict, Any
from pathlib import Path
import logging

logger = logging.getLogger(__name__)

class CharacterParser:
    def __init__(self, data_directory: str = "data/characters"):
        self.data_directory = Path(data_directory)
        self.section_patterns = [
            r'^[A-ZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀĀĂĄĆČĎĐĒĔĖĘĚĜĞĠĢĤĦĪĬĮİĲĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŪŬŮŰŲŴŶŸŹŻŽ\s]+$',  # All caps line
            r'^[A-ZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀĀĂĄĆČĎĐĒĔĖĘĚĜĞĠĢĤĦĪĬĮİĲĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŪŬŮŰŲŴŶŸŹŻŽ\s].*[A-ZÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀĀĂĄĆČĎĐĒĔĖĘĚĜĞĠĢĤĦĪĬĮİĲĴĶĹĻĽĿŁŃŅŇŊŌŎŐŒŔŖŘŚŜŞŠŢŤŦŪŬŮŰŲŴŶŸŹŻŽ]$',  # Mostly caps
        ]
        
        # Decorative line patterns
        self.decorative_patterns = [
            r'^={3,}',  # ===
            r'^-{3,}',  # ---
            r'^_{3,}',  # ___
            r'^#{3,}',  # ###
            r'^\*{3,}', # ***
        ]

    def is_section_header(self, line: str) -> bool:
        """Check if a line is a section header"""
        line = line.strip()
        if not line:
            return False
            
        # Check for decorative patterns
        for pattern in self.decorative_patterns:
            if re.match(pattern, line):
                return False
                
        # Check for section patterns
        for pattern in self.section_patterns:
            if re.match(pattern, line, re.IGNORECASE):
                return True
                
        return False

    def is_decorative_line(self, line: str) -> bool:
        """Check if a line is decorative (separators)"""
        line = line.strip()
        for pattern in self.decorative_patterns:
            if re.match(pattern, line):
                return True
        return False

    def extract_character_name(self, content: str) -> str:
        """Extract character name from content"""
        lines = content.split('\n')
        
        for line in lines:
            line = line.strip()
            
            # Look for "NOME DO PERSONAGEM:" pattern
            if "NOME DO PERSONAGEM:" in line.upper():
                name_part = line.split(":", 1)
                if len(name_part) > 1:
                    return name_part[1].strip()
            
            # Look for "PERFIL DE PERSONAGEM -" pattern
            if "PERFIL DE PERSONAGEM -" in line.upper():
                name_part = line.split("-", 1)
                if len(name_part) > 1:
                    return name_part[1].strip()
        
        return "Nome não encontrado"

    def parse_content_into_sections(self, content: str) -> List[Dict[str, Any]]:
        """Parse content into sections intelligently"""
        lines = content.split('\n')
        sections = []
        current_section = None
        current_items = []
        
        for i, line in enumerate(lines):
            line = line.strip()
            
            # Skip empty lines and decorative lines
            if not line or self.is_decorative_line(line):
                continue
            
            # Check if it's a section header
            if self.is_section_header(line):
                # Save previous section if exists
                if current_section and current_items:
                    sections.append({
                        "title": current_section,
                        "content": current_items
                    })
                
                # Start new section
                current_section = line
                current_items = []
                continue
            
            # Process content within section
            if current_section:
                # Check if line contains a colon (key-value pair)
                if ':' in line and not line.startswith('-') and not line.startswith('*'):
                    parts = line.split(':', 1)
                    if len(parts) == 2:
                        key = parts[0].strip()
                        value = parts[1].strip()
                        if key and value:
                            current_items.append({
                                "subtitle": key,
                                "text": value
                            })
                            continue
                
                # Check if it's a numbered or bulleted list item
                if re.match(r'^[\d\-\*]\s*\.?\s*', line):
                    # Remove bullet/number and clean up
                    cleaned_line = re.sub(r'^[\d\-\*]\s*\.?\s*', '', line).strip()
                    if cleaned_line:
                        current_items.append({
                            "text": cleaned_line
                        })
                    continue
                
                # Check if it's a sub-item with dash
                if line.startswith('-') or line.startswith('*'):
                    cleaned_line = line[1:].strip()
                    if cleaned_line:
                        current_items.append({
                            "text": cleaned_line
                        })
                    continue
                
                # Regular paragraph
                if line and len(line) > 10:  # Ignore very short lines
                    current_items.append({
                        "text": line
                    })
        
        # Don't forget the last section
        if current_section and current_items:
            sections.append({
                "title": current_section,
                "content": current_items
            })
        
        return sections

    def parse_character_file(self, file_path: Path) -> Dict[str, Any]:
        """Parse a single character file"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            character_name = self.extract_character_name(content)
            sections = self.parse_content_into_sections(content)
            
            # Extract basic info from first section if available
            title = "Personagem"
            if sections and sections[0]["title"].upper() in ["INFORMAÇÕES BÁSICAS", "INFORMACOES BASICAS"]:
                for item in sections[0]["content"]:
                    if item.get("subtitle", "").lower() == "tipo":
                        title = item["text"]
                        break
            
            return {
                "id": file_path.stem,
                "name": character_name,
                "title": title,
                "sections": sections,
                "file_path": str(file_path)
            }
            
        except Exception as e:
            logger.error(f"Error parsing file {file_path}: {str(e)}")
            return None

    def get_all_characters(self) -> List[Dict[str, Any]]:
        """Get all characters from the data directory"""
        characters = []
        
        if not self.data_directory.exists():
            logger.warning(f"Data directory {self.data_directory} does not exist")
            return characters
        
        # Get all txt files
        txt_files = list(self.data_directory.glob("*.txt"))
        
        for file_path in txt_files:
            character = self.parse_character_file(file_path)
            if character:
                characters.append(character)
        
        # Sort by name
        characters.sort(key=lambda x: x["name"])
        
        return characters

    def get_character_by_id(self, character_id: str) -> Dict[str, Any]:
        """Get a specific character by ID"""
        file_path = self.data_directory / f"{character_id}.txt"
        
        if not file_path.exists():
            return None
        
        return self.parse_character_file(file_path)

    def search_characters(self, query: str) -> List[Dict[str, Any]]:
        """Search characters by name or content"""
        all_characters = self.get_all_characters()
        
        if not query:
            return all_characters
        
        query_lower = query.lower()
        filtered_characters = []
        
        for character in all_characters:
            # Search in name and title
            if (query_lower in character["name"].lower() or 
                query_lower in character["title"].lower()):
                filtered_characters.append(character)
                continue
            
            # Search in section content
            found_in_content = False
            for section in character["sections"]:
                if query_lower in section["title"].lower():
                    found_in_content = True
                    break
                for item in section["content"]:
                    if query_lower in item.get("text", "").lower():
                        found_in_content = True
                        break
                    if query_lower in item.get("subtitle", "").lower():
                        found_in_content = True
                        break
                if found_in_content:
                    break
            
            if found_in_content:
                filtered_characters.append(character)
        
        return filtered_characters